File: src/CircleAI.Domain/InMemoryDomain.cs
// InMemoryDomain.cs
//
// (3.3.0) Real-but-lightweight in-memory implementations for every
// CircleAI.Domain contract surface. None of these are LLM-grade
// experts — they are the deterministic in-process backings tests
// expect, and the fallbacks production hosts swap out one-by-one as
// real specialists (EPICure / quant-mind / dexter / presenton /
// career-ops / mempalace / HippoRAG / MiroFish) get vendored.

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace CircleAI.Domain;

// ─── Food (substitute-by-canonical-name) ───────────────────────────────
public sealed class InMemoryFoodEmbeddings : IFoodEmbeddings
{
    private readonly ConcurrentDictionary<string, float[]> _embeds = new(StringComparer.OrdinalIgnoreCase);
    private readonly ConcurrentDictionary<string, List<Ingredient>> _subs = new(StringComparer.OrdinalIgnoreCase);

    public string BackendId => "in-memory";

    public void RegisterEmbedding(string name, float[] v) => _embeds[name] = v ?? throw new ArgumentNullException(nameof(v));
    public void RegisterSubstitute(string name, Ingredient alt)
    {
        var list = _subs.GetOrAdd(name, _ => new List<Ingredient>());
        list.Add(alt ?? throw new ArgumentNullException(nameof(alt)));
    }

    public ValueTask<float[]> EmbedAsync(Ingredient i, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(i);
        if (_embeds.TryGetValue(i.Name, out var v)) return ValueTask.FromResult(v);
        // Deterministic hash-based 8-dim vector if no embedding was registered.
        var v2 = new float[8];
        var h = i.Name.GetHashCode(StringComparison.OrdinalIgnoreCase);
        for (var k = 0; k < 8; k++) v2[k] = ((h >> (k * 4)) & 0xF) / 15f;
        return ValueTask.FromResult(v2);
    }

    public ValueTask<IReadOnlyList<Ingredient>> SubstitutesAsync(Ingredient i, int topK = 5, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(i);
        if (topK <= 0) throw new ArgumentOutOfRangeException(nameof(topK));
        if (!_subs.TryGetValue(i.Name, out var list)) return ValueTask.FromResult<IReadOnlyList<Ingredient>>(Array.Empty<Ingredient>());
        return ValueTask.FromResult<IReadOnlyList<Ingredient>>(list.Take(topK).ToArray());
    }
}

// ─── Finance ───────────────────────────────────────────────────────────
public sealed class InMemoryFinanceRetrieval : IFinanceRetrieval
{
    private readonly List<FinanceSnippet> _corpus = new();
    private readonly object _lock = new();
    public string BackendId => "in-memory";

    public void Add(FinanceSnippet s) { ArgumentNullException.ThrowIfNull(s); lock (_lock) _corpus.Add(s); }

    public ValueTask<IReadOnlyList<FinanceSnippet>> RetrieveAsync(string query, int topK = 5, CancellationToken ct = default)
    {
        if (query is null) throw new ArgumentNullException(nameof(query));
        if (topK <= 0)     throw new ArgumentOutOfRangeException(nameof(topK));
        lock (_lock)
        {
            return ValueTask.FromResult<IReadOnlyList<FinanceSnippet>>(
                _corpus.Where(s => s.Text.Contains(query, StringComparison.OrdinalIgnoreCase))
                       .OrderByDescending(s => s.Score)
                       .Take(topK).ToArray());
        }
    }
}

/// <summary>(3.3.0) Real financial agent — runs multi-pass retrieval (decomposes the
/// question into sub-questions), groups findings by source, summarises each cluster.</summary>
public sealed class MultiPassFinancialAgent : IFinancialAgent
{
    private readonly IFinanceRetrieval _retr;
    public MultiPassFinancialAgent(IFinanceRetrieval r) => _retr = r ?? throw new ArgumentNullException(nameof(r));
    public string BackendId => "multi-pass";

    public async ValueTask<IReadOnlyList<FinanceFinding>> ResearchAsync(string question, CancellationToken ct = default)
    {
        if (question is null) throw new ArgumentNullException(nameof(question));
        var subQuestions = Decompose(question);
        var findings = new List<FinanceFinding>();
        foreach (var sub in subQuestions)
        {
            var snippets = await _retr.RetrieveAsync(sub, 5, ct).ConfigureAwait(false);
            if (snippets.Count == 0) continue;
            var bySource = snippets.GroupBy(s => s.Source);
            foreach (var grp in bySource)
            {
                var summary = string.Join(" | ", grp.OrderByDescending(s => s.Score).Take(3).Select(s => s.Text));
                findings.Add(new FinanceFinding(Subject: sub, Summary: summary, Citations: new[] { grp.Key }));
            }
        }
        return findings.Count == 0 ? Array.Empty<FinanceFinding>() : findings;
    }

    private static IReadOnlyList<string> Decompose(string question)
    {
        var subs = new List<string> { question };
        if (question.Contains(" and ", StringComparison.OrdinalIgnoreCase))
            foreach (var part in question.Split(new[] { " and " }, StringSplitOptions.RemoveEmptyEntries))
                if (part.Trim().Length > 6) subs.Add(part.Trim());
        if (question.Length > 60)
        {
            subs.Add(question.Split(',').First().Trim());
        }
        return subs.Distinct().ToArray();
    }
}


// ─── Presentations ──────────────────────────────────────────────────────
public sealed class TemplatePresentationGenerator : IPresentationGenerator
{
    public string BackendId => "template";

    public ValueTask<GeneratedPresentation> GenerateAsync(string topic, int targetSlideCount = 10, string? theme = null, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(topic))  throw new ArgumentException("topic required");
        if (targetSlideCount <= 0)             throw new ArgumentOutOfRangeException(nameof(targetSlideCount));
        var slides = new List<SlideOutline>(targetSlideCount);
        slides.Add(new SlideOutline(topic, "Overview", new[] { "What is " + topic, "Why it matters", "What we'll cover" }));
        for (var i = 2; i < targetSlideCount; i++)
        {
            slides.Add(new SlideOutline($"{topic} — Part {i - 1}", $"Detail for part {i - 1}", new[] { "Point A", "Point B", "Point C" }));
        }
        slides.Add(new SlideOutline("Conclusion", $"Summary of {topic}", new[] { "Recap", "Next steps", "Questions" }));
        return ValueTask.FromResult(new GeneratedPresentation(slides, theme ?? "default", "markdown"));
    }
}

// ─── Job search ─────────────────────────────────────────────────────────
public sealed class TemplateJobSearchPipeline : IJobSearchPipeline
{
    public string BackendId => "template";

    public ValueTask<JobApplicationDraft> DraftApplicationAsync(string roleDescription, string candidateProfileText, CancellationToken ct = default)
    {
        if (roleDescription is null)      throw new ArgumentNullException(nameof(roleDescription));
        if (candidateProfileText is null) throw new ArgumentNullException(nameof(candidateProfileText));
        var roleWords = ExtractKeyWords(roleDescription);
        var candWords = ExtractKeyWords(candidateProfileText);
        var matches = roleWords.Intersect(candWords, StringComparer.OrdinalIgnoreCase).Take(10).ToArray();
        var resume = $"{candidateProfileText.Trim()}\n\nMatched skills: {string.Join(", ", matches)}";
        var cover  = $"Dear Hiring Team,\n\nI am applying because my background ({string.Join(", ", matches.Take(3))}) fits the role.\n\nRegards.";
        return ValueTask.FromResult(new JobApplicationDraft(resume, cover, matches));
    }

    private static IReadOnlyList<string> ExtractKeyWords(string text)
        => text.Split(new[] { ' ', '\n', '\r', '\t', ',', '.', ';', ':', '(', ')' }, StringSplitOptions.RemoveEmptyEntries)
               .Where(w => w.Length > 3).Select(w => w.Trim().ToLowerInvariant()).Distinct().ToArray();
}

// ─── Memory upgrades ────────────────────────────────────────────────────
public sealed class InMemoryMemPalaceStore : IMemPalaceStore
{
    private readonly ConcurrentDictionary<string, MemoryItem> _items = new(StringComparer.Ordinal);
    public string BackendId => "in-memory";

    public ValueTask UpsertAsync(MemoryItem item, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(item);
        if (string.IsNullOrWhiteSpace(item.Id)) throw new ArgumentException("Id required");
        _items[item.Id] = item;
        return ValueTask.CompletedTask;
    }

    public ValueTask<IReadOnlyList<MemoryHit>> RecallAsync(string query, int topK = 5, CancellationToken ct = default)
    {
        if (query is null) throw new ArgumentNullException(nameof(query));
        if (topK <= 0)     throw new ArgumentOutOfRangeException(nameof(topK));
        var hits = _items.Values
            .Select(i => new MemoryHit(i, Score(i.Text, query)))
            .Where(h => h.Score > 0)
            .OrderByDescending(h => h.Score)
            .Take(topK)
            .ToArray();
        return ValueTask.FromResult<IReadOnlyList<MemoryHit>>(hits);
    }

    private static float Score(string body, string query)
    {
        if (string.IsNullOrEmpty(body) || string.IsNullOrEmpty(query)) return 0f;
        var q = query.Trim();
        var idx = body.IndexOf(q, StringComparison.OrdinalIgnoreCase);
        return idx < 0 ? 0f : 1f / (1f + idx);
    }
}

public sealed class InMemoryHippoRagStore : IHippoRagStore
{
    private readonly InMemoryMemPalaceStore _base = new();
    public string BackendId => "in-memory";

    public ValueTask IndexAsync(MemoryItem item, CancellationToken ct = default) => _base.UpsertAsync(item, ct);

    public async ValueTask<IReadOnlyList<MemoryHit>> MultiHopRecallAsync(string query, int topK = 5, CancellationToken ct = default)
    {
        // Multi-hop: first hop with the query, then expand using top hit's text as a follow-up query.
        var first = await _base.RecallAsync(query, topK, ct).ConfigureAwait(false);
        if (first.Count == 0) return first;
        var seed   = first[0].Item.Text;
        var second = await _base.RecallAsync(seed, topK, ct).ConfigureAwait(false);
        return first.Concat(second).GroupBy(h => h.Item.Id).Select(g => g.First())
                    .OrderByDescending(h => h.Score).Take(topK).ToArray();
    }
}

// ─── Swarm ──────────────────────────────────────────────────────────────
public sealed class InMemorySwarmCoordinator : ISwarmCoordinator
{
    private readonly ConcurrentDictionary<string, SwarmPeer> _peers = new(StringComparer.Ordinal);
    public string BackendId => "in-memory";

    public void Register(SwarmPeer p) { ArgumentNullException.ThrowIfNull(p); _peers[p.PeerId] = p; }

    public ValueTask<IReadOnlyList<SwarmPeer>> ListPeersAsync(CancellationToken ct = default)
        => ValueTask.FromResult<IReadOnlyList<SwarmPeer>>(_peers.Values.ToArray());

    public ValueTask<string?> ChooseDelegateAsync(string capability, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(capability)) throw new ArgumentException("capability required", nameof(capability));
        var pick = _peers.Values
            .Where(p => string.Equals(p.Capability, capability, StringComparison.OrdinalIgnoreCase))
            .OrderByDescending(p => p.Health)
            .FirstOrDefault();
        return ValueTask.FromResult(pick?.PeerId);
    }
}

// ─── Personal LoRA — real in-memory adapter manager with a simulated training loop ─────
public sealed class InMemoryPersonalLoRA : IPersonalLoRA
{
    private readonly ConcurrentDictionary<string, LoRAAdapterState> _adapters = new(StringComparer.Ordinal);
    private readonly ConcurrentDictionary<string, byte> _loaded = new(StringComparer.Ordinal);

    public string BackendId => "in-memory";

    public ValueTask<LoRATrainingSummary> TrainAsync(string adapterId, IReadOnlyList<string> samples, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(adapterId)) throw new ArgumentException("adapterId required");
        ArgumentNullException.ThrowIfNull(samples);
        if (samples.Count == 0) throw new ArgumentException("at least one sample required");

        // Simulated training loop: each sample contributes a step. Final loss
        // decreases logarithmically with sample count (a realistic baseline shape).
        var steps     = samples.Count;
        var totalChars = samples.Sum(s => s?.Length ?? 0);
        var finalLoss = (float)(1.0 / (1.0 + Math.Log(1 + steps)) + 1.0 / (1.0 + totalChars / 1000.0));
        var state = new LoRAAdapterState(adapterId, steps, finalLoss, DateTimeOffset.UtcNow);
        _adapters[adapterId] = state;
        return ValueTask.FromResult(new LoRATrainingSummary(adapterId, steps, finalLoss));
    }

    public ValueTask LoadAdapterAsync(string adapterId, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(adapterId)) throw new ArgumentException("adapterId required");
        if (!_adapters.ContainsKey(adapterId))
            throw new InvalidOperationException($"Adapter '{adapterId}' not trained.");
        _loaded[adapterId] = 1;
        return ValueTask.CompletedTask;
    }

    public ValueTask UnloadAdapterAsync(string adapterId, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(adapterId)) throw new ArgumentException("adapterId required");
        _loaded.TryRemove(adapterId, out _);
        return ValueTask.CompletedTask;
    }

    public bool IsLoaded(string adapterId) => _loaded.ContainsKey(adapterId);
    public LoRAAdapterState? StateOf(string adapterId) => _adapters.GetValueOrDefault(adapterId);
}

public sealed record LoRAAdapterState(string AdapterId, int Steps, float FinalLoss, DateTimeOffset TrainedAtUtc);

File: python/src/circle_ai/domain/in_memory_domain.py
# in_memory_domain.py
#
# Port of CircleAI.Domain InMemoryDomain.cs (C# — the EXACT spec).
#
# (3.3.0) Real-but-lightweight in-memory backings for every CircleAI.Domain
# contract — the deterministic in-process fallbacks production hosts swap out
# one-by-one as real specialists get vendored.
#
# Notes on faithful porting:
#   • InMemoryFoodEmbeddings.embed_async: when no embedding is registered the C#
#     builds an 8-dim vector from `name.GetHashCode(OrdinalIgnoreCase)` nibbles.
#     .NET's string hash is randomised per-process (not portable), so this port
#     uses a fixed FNV-1a over the lower-cased name — same shape (8 floats in
#     [0,1] from 4-bit nibbles), deterministic across runs. float32 at the /15f.
#   • MultiPassFinancialAgent: decompose on " and " + long-question leading
#     clause, dedup preserving order, retrieve per sub-question, group by source,
#     summarise the top-3-by-score snippets " | "-joined.
#   • InMemoryPersonalLoRA: simulated training loss
#     1/(1+ln(1+steps)) + 1/(1+totalChars/1000) as float32.
#   • dict iteration order == insertion order in both CPython and C#
#     ConcurrentDictionary snapshots for these read-all paths.

from __future__ import annotations

import math
import struct
import threading
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Dict, List, Optional, Sequence

from .contracts import (
    FinanceFinding,
    FinanceSnippet,
    GeneratedPresentation,
    IFinancialAgent,
    IFinanceRetrieval,
    IFoodEmbeddings,
    IHippoRagStore,
    IJobSearchPipeline,
    IMemPalaceStore,
    IPersonalLoRA,
    IPresentationGenerator,
    ISwarmCoordinator,
    Ingredient,
    JobApplicationDraft,
    LoRATrainingSummary,
    MemoryHit,
    MemoryItem,
    SlideOutline,
    SwarmPeer,
)


def _f32(x: float) -> float:
    return struct.unpack("<f", struct.pack("<f", x))[0]


def _stable_hash(text: str) -> int:
    # Deterministic 32-bit FNV-1a over the lower-cased text (stand-in for the
    # per-process-randomised .NET string GetHashCode used by the C# fallback).
    h = 2166136261
    for b in text.lower().encode("utf-8"):
        h ^= b
        h = (h * 16777619) & 0xFFFFFFFF
    return h


# ─── Food (substitute-by-canonical-name) ───────────────────────────────
class InMemoryFoodEmbeddings(IFoodEmbeddings):
    """Real in-memory :class:`IFoodEmbeddings`. Mirrors
    ``CircleAI.Domain.InMemoryFoodEmbeddings``."""

    def __init__(self) -> None:
        self._embeds: Dict[str, List[float]] = {}
        self._subs: Dict[str, List[Ingredient]] = {}
        self._lock = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    def register_embedding(self, name: str, v: List[float]) -> None:
        if v is None:
            raise ValueError("v")
        with self._lock:
            self._embeds[name.lower()] = list(v)

    def register_substitute(self, name: str, alt: Ingredient) -> None:
        if alt is None:
            raise ValueError("alt")
        with self._lock:
            self._subs.setdefault(name.lower(), []).append(alt)

    async def embed_async(
        self, ingredient: Ingredient, ct: Optional[object] = None
    ) -> List[float]:
        if ingredient is None:
            raise ValueError("i")
        with self._lock:
            got = self._embeds.get(ingredient.name.lower())
        if got is not None:
            return list(got)
        # Deterministic hash-based 8-dim vector if no embedding was registered.
        v2 = [0.0] * 8
        h = _stable_hash(ingredient.name)
        for k in range(8):
            v2[k] = _f32(((h >> (k * 4)) & 0xF) / 15.0)
        return v2

    async def substitutes_async(
        self, ingredient: Ingredient, top_k: int = 5, ct: Optional[object] = None
    ) -> List[Ingredient]:
        if ingredient is None:
            raise ValueError("i")
        if top_k <= 0:
            raise ValueError("topK")
        with self._lock:
            got = self._subs.get(ingredient.name.lower())
            if got is None:
                return []
            return list(got[:top_k])


# ─── Finance ───────────────────────────────────────────────────────────
class InMemoryFinanceRetrieval(IFinanceRetrieval):
    """Real in-memory :class:`IFinanceRetrieval`. Mirrors
    ``CircleAI.Domain.InMemoryFinanceRetrieval``."""

    def __init__(self) -> None:
        self._corpus: List[FinanceSnippet] = []
        self._lock = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    def add(self, s: FinanceSnippet) -> None:
        if s is None:
            raise ValueError("s")
        with self._lock:
            self._corpus.append(s)

    async def retrieve_async(
        self, query: str, top_k: int = 5, ct: Optional[object] = None
    ) -> List[FinanceSnippet]:
        if query is None:
            raise ValueError("query")
        if top_k <= 0:
            raise ValueError("topK")
        ql = query.lower()
        with self._lock:
            hits = [s for s in self._corpus if ql in s.text.lower()]
        hits.sort(key=lambda s: s.score, reverse=True)
        return hits[:top_k]


class MultiPassFinancialAgent(IFinancialAgent):
    """Real financial agent — multi-pass retrieval + per-source summaries.
    Mirrors ``CircleAI.Domain.MultiPassFinancialAgent``."""

    def __init__(self, r: IFinanceRetrieval) -> None:
        if r is None:
            raise ValueError("r")
        self._retr = r

    @property
    def backend_id(self) -> str:
        return "multi-pass"

    async def research_async(
        self, question: str, ct: Optional[object] = None
    ) -> List[FinanceFinding]:
        if question is None:
            raise ValueError("question")
        sub_questions = self._decompose(question)
        findings: List[FinanceFinding] = []
        for sub in sub_questions:
            snippets = await self._retr.retrieve_async(sub, 5, ct)
            if len(snippets) == 0:
                continue
            # GroupBy(Source) — preserve first-seen source order.
            by_source: Dict[str, List[FinanceSnippet]] = {}
            for s in snippets:
                by_source.setdefault(s.source, []).append(s)
            for source, grp in by_source.items():
                top3 = sorted(grp, key=lambda s: s.score, reverse=True)[:3]
                summary = " | ".join(s.text for s in top3)
                findings.append(
                    FinanceFinding(subject=sub, summary=summary, citations=[source])
                )
        return findings

    @staticmethod
    def _decompose(question: str) -> List[str]:
        subs: List[str] = [question]
        if " and " in question.lower():
            # Split case-insensitively on " and " (mirror the C# split on the
            # literal " and " token; .NET Split is ordinal here).
            for part in _split_ci(question, " and "):
                if len(part.strip()) > 6:
                    subs.append(part.strip())
        if len(question) > 60:
            subs.append(question.split(",")[0].strip())
        # Distinct preserving order.
        seen = set()
        out: List[str] = []
        for s in subs:
            if s not in seen:
                seen.add(s)
                out.append(s)
        return out


def _split_ci(text: str, sep: str) -> List[str]:
    # Case-insensitive split on `sep`, RemoveEmptyEntries.
    parts: List[str] = []
    low = text.lower()
    seplow = sep.lower()
    start = 0
    while True:
        idx = low.find(seplow, start)
        if idx < 0:
            piece = text[start:]
            if piece:
                parts.append(piece)
            break
        piece = text[start:idx]
        if piece:
            parts.append(piece)
        start = idx + len(sep)
    return parts


# ─── Presentations ──────────────────────────────────────────────────────
class TemplatePresentationGenerator(IPresentationGenerator):
    """Template presentation generator. Mirrors
    ``CircleAI.Domain.TemplatePresentationGenerator``."""

    @property
    def backend_id(self) -> str:
        return "template"

    async def generate_async(
        self,
        topic: str,
        target_slide_count: int = 10,
        theme: Optional[str] = None,
        ct: Optional[object] = None,
    ) -> GeneratedPresentation:
        if topic is None or topic.strip() == "":
            raise ValueError("topic required")
        if target_slide_count <= 0:
            raise ValueError("targetSlideCount")
        slides: List[SlideOutline] = []
        slides.append(
            SlideOutline(
                topic,
                "Overview",
                ["What is " + topic, "Why it matters", "What we'll cover"],
            )
        )
        for i in range(2, target_slide_count):
            slides.append(
                SlideOutline(
                    f"{topic} — Part {i - 1}",
                    f"Detail for part {i - 1}",
                    ["Point A", "Point B", "Point C"],
                )
            )
        slides.append(
            SlideOutline(
                "Conclusion", f"Summary of {topic}", ["Recap", "Next steps", "Questions"]
            )
        )
        return GeneratedPresentation(slides, theme if theme is not None else "default", "markdown")


# ─── Job search ─────────────────────────────────────────────────────────
class TemplateJobSearchPipeline(IJobSearchPipeline):
    """Template job-search pipeline. Mirrors
    ``CircleAI.Domain.TemplateJobSearchPipeline``."""

    @property
    def backend_id(self) -> str:
        return "template"

    async def draft_application_async(
        self,
        role_description: str,
        candidate_profile_text: str,
        ct: Optional[object] = None,
    ) -> JobApplicationDraft:
        if role_description is None:
            raise ValueError("roleDescription")
        if candidate_profile_text is None:
            raise ValueError("candidateProfileText")
        role_words = self._extract_key_words(role_description)
        cand_words = set(self._extract_key_words(candidate_profile_text))
        # Intersect preserving role-word order, take 10.
        matches: List[str] = []
        for w in role_words:
            if w in cand_words and w not in matches:
                matches.append(w)
            if len(matches) == 10:
                break
        resume = f"{candidate_profile_text.strip()}\n\nMatched skills: {', '.join(matches)}"
        cover = (
            "Dear Hiring Team,\n\nI am applying because my background "
            f"({', '.join(matches[:3])}) fits the role.\n\nRegards."
        )
        return JobApplicationDraft(resume, cover, matches)

    @staticmethod
    def _extract_key_words(text: str) -> List[str]:
        seps = [" ", "\n", "\r", "\t", ",", ".", ";", ":", "(", ")"]
        raw = [text]
        for sep in seps:
            nxt: List[str] = []
            for chunk in raw:
                nxt.extend(chunk.split(sep))
            raw = nxt
        out: List[str] = []
        seen = set()
        for w in raw:
            if len(w) > 3:
                lw = w.strip().lower()
                if lw not in seen:
                    seen.add(lw)
                    out.append(lw)
        return out


# ─── Memory upgrades ────────────────────────────────────────────────────
class InMemoryMemPalaceStore(IMemPalaceStore):
    """Real in-memory :class:`IMemPalaceStore`. Mirrors
    ``CircleAI.Domain.InMemoryMemPalaceStore``."""

    def __init__(self) -> None:
        self._items: Dict[str, MemoryItem] = {}
        self._lock = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    async def upsert_async(
        self, item: MemoryItem, ct: Optional[object] = None
    ) -> None:
        if item is None:
            raise ValueError("item")
        if item.id is None or item.id.strip() == "":
            raise ValueError("Id required")
        with self._lock:
            self._items[item.id] = item

    async def recall_async(
        self, query: str, top_k: int = 5, ct: Optional[object] = None
    ) -> List[MemoryHit]:
        if query is None:
            raise ValueError("query")
        if top_k <= 0:
            raise ValueError("topK")
        with self._lock:
            items = list(self._items.values())
        hits = [MemoryHit(i, self._score(i.text, query)) for i in items]
        hits = [h for h in hits if h.score > 0]
        hits.sort(key=lambda h: h.score, reverse=True)
        return hits[:top_k]

    @staticmethod
    def _score(body: str, query: str) -> float:
        if not body or not query:
            return 0.0
        q = query.strip()
        idx = body.lower().find(q.lower())
        return 0.0 if idx < 0 else _f32(1.0 / (1.0 + idx))


class InMemoryHippoRagStore(IHippoRagStore):
    """Real in-memory :class:`IHippoRagStore` (multi-hop over a MemPalace).
    Mirrors ``CircleAI.Domain.InMemoryHippoRagStore``."""

    def __init__(self) -> None:
        self._base = InMemoryMemPalaceStore()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    async def index_async(
        self, item: MemoryItem, ct: Optional[object] = None
    ) -> None:
        await self._base.upsert_async(item, ct)

    async def multi_hop_recall_async(
        self, query: str, top_k: int = 5, ct: Optional[object] = None
    ) -> List[MemoryHit]:
        first = await self._base.recall_async(query, top_k, ct)
        if len(first) == 0:
            return first
        seed = first[0].item.text
        second = await self._base.recall_async(seed, top_k, ct)
        # Concat, GroupBy(Item.Id).First(), OrderByDescending(Score), Take(topK).
        merged: Dict[str, MemoryHit] = {}
        for h in list(first) + list(second):
            if h.item.id not in merged:
                merged[h.item.id] = h
        out = list(merged.values())
        out.sort(key=lambda h: h.score, reverse=True)
        return out[:top_k]


# ─── Swarm ──────────────────────────────────────────────────────────────
class InMemorySwarmCoordinator(ISwarmCoordinator):
    """Real in-memory :class:`ISwarmCoordinator`. Mirrors
    ``CircleAI.Domain.InMemorySwarmCoordinator``."""

    def __init__(self) -> None:
        self._peers: Dict[str, SwarmPeer] = {}
        self._lock = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    def register(self, p: SwarmPeer) -> None:
        if p is None:
            raise ValueError("p")
        with self._lock:
            self._peers[p.peer_id] = p

    async def list_peers_async(
        self, ct: Optional[object] = None
    ) -> List[SwarmPeer]:
        with self._lock:
            return list(self._peers.values())

    async def choose_delegate_async(
        self, capability: str, ct: Optional[object] = None
    ) -> Optional[str]:
        if capability is None or capability.strip() == "":
            raise ValueError("capability required")
        cl = capability.lower()
        with self._lock:
            candidates = [p for p in self._peers.values() if p.capability.lower() == cl]
        if not candidates:
            return None
        # OrderByDescending(Health).FirstOrDefault() — stable; keep first on ties.
        pick = max(candidates, key=lambda p: p.health)
        # max() returns the first max on ties (matches OrderByDescending+First).
        best_health = pick.health
        for p in candidates:
            if p.health == best_health:
                return p.peer_id
        return pick.peer_id


# ─── Personal LoRA ──────────────────────────────────────────────────────
@dataclass(frozen=True, slots=True)
class LoRAAdapterState:
    """Mirrors ``CircleAI.Domain.LoRAAdapterState`` — ``record(string AdapterId,
    int Steps, float FinalLoss, DateTimeOffset TrainedAtUtc)``.
    """

    adapter_id: str
    steps: int
    final_loss: float
    trained_at_utc: datetime


class InMemoryPersonalLoRA(IPersonalLoRA):
    """Real in-memory :class:`IPersonalLoRA` with a simulated training loop.
    Mirrors ``CircleAI.Domain.InMemoryPersonalLoRA``."""

    def __init__(self) -> None:
        self._adapters: Dict[str, LoRAAdapterState] = {}
        self._loaded: Dict[str, int] = {}
        self._lock = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    async def train_async(
        self,
        adapter_id: str,
        conversation_samples: Sequence[str],
        ct: Optional[object] = None,
    ) -> LoRATrainingSummary:
        if adapter_id is None or adapter_id.strip() == "":
            raise ValueError("adapterId required")
        if conversation_samples is None:
            raise ValueError("samples")
        samples = list(conversation_samples)
        if len(samples) == 0:
            raise ValueError("at least one sample required")
        steps = len(samples)
        total_chars = sum(len(s) if s is not None else 0 for s in samples)
        final_loss = _f32(
            1.0 / (1.0 + math.log(1 + steps)) + 1.0 / (1.0 + total_chars / 1000.0)
        )
        state = LoRAAdapterState(adapter_id, steps, final_loss, datetime.now(timezone.utc))
        with self._lock:
            self._adapters[adapter_id] = state
        return LoRATrainingSummary(adapter_id, steps, final_loss)

    async def load_adapter_async(
        self, adapter_id: str, ct: Optional[object] = None
    ) -> None:
        if adapter_id is None or adapter_id.strip() == "":
            raise ValueError("adapterId required")
        with self._lock:
            if adapter_id not in self._adapters:
                raise RuntimeError(f"Adapter '{adapter_id}' not trained.")
            self._loaded[adapter_id] = 1

    async def unload_adapter_async(
        self, adapter_id: str, ct: Optional[object] = None
    ) -> None:
        if adapter_id is None or adapter_id.strip() == "":
            raise ValueError("adapterId required")
        with self._lock:
            self._loaded.pop(adapter_id, None)

    def is_loaded(self, adapter_id: str) -> bool:
        with self._lock:
            return adapter_id in self._loaded

    def state_of(self, adapter_id: str) -> Optional[LoRAAdapterState]:
        with self._lock:
            return self._adapters.get(adapter_id)
